Added Destructible script

j-tomada opened the pull request:
Changes object's texture at certain health percentages.

Inherits from entity class

j-tomada committed:
Added Destructible script

Changes object's texture at certain health percentages.

Inherits from entity class

j-tomada committed:
Added new lines for optimization

the ChangeTextures method should now only run if the object has lost health.

Removed files that were used for testing.

BUNN1E5 commented on Assets/Scripts/Destructible.cs:
```diff
@@ -0,0 +1,82 @@
+        MaxHealth = base.health;
+        prevHealth = base.health;
+        HealthPercents = CreatePercents(Sprites.Length);
+        gameObject.GetComponent<SpriteRenderer>().sprite = Sprites[0];
```
You call gameObject.GetComponent<SpriteRenderer>().sprite more than once, you should cache it in a local varaible

j-tomada replied:
Might just be overcomplicating it, but I'm not exactly sure how to do it. Any tips?

BUNN1E5 replied:
You can just create a global SpriteRenderer then assign it `spriteRenderer = gameObject.GetComponent<SpriteRenderer>()` then when you want to change the sprite, just do `spriteRenderer.sprite = Sprites[0];`

BUNN1E5 commented on Assets/Scripts/Destructible.cs:
```diff
@@ -0,0 +1,82 @@
+
+            if (currPercent < HealthPercents[i] & currPercent >= HealthPercents[i + 1])
+            {
+                gameObject.GetComponent<SpriteRenderer>().sprite = Sprites[i + 1];
```
So I'm not sure what exactly is happening here. Can you explain to me the logic here?

Please correct me if I'm wrong, but it seems like you are potentially making a new sprite for each health point?

j-tomada replied:
So the HealthPercents array basically highlights the cutoffs at which the object should change textures (Array would typically look like: 100, 75, 50, 25, 0). If currPercent, which is the object's current health as a percentage, is within a certain range (Like between 75 - 50), then it will change to the next sprite. Else, it will just stay as as its current Sprite texture.

BUNN1E5 replied:
Ah I see, so these are like different damage sprites? So as the Destructable gets more damaged?

BUNN1E5 commented on Assets/Scripts/Destructible.cs:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    //Overrides the OnDeath method inherited from the Entity class
+    public override void OnDeath()
```
Not really needed the Entity class has this exact code lol

Kainkun replied:
Yeah make sure to read whats in the entity class that you inherit from. that should help you understand whats going on

BUNN1E5 commented on Assets/Scripts/Destructible.cs:
```diff
@@ -0,0 +1,82 @@
+    {
+        if (base.health <= 0)
+        {
+            OnDeath();
```
This is fine _for now_ but keep in mind that the entity script should be handling this instead.

BUNN1E5 commented on Assets/Scripts/Destructible.cs:
```diff
@@ -0,0 +1,82 @@
+
+        if (base.health < prevHealth) //Now will only update if Object has lost health
+        {
+            ChangeTextures();
```
Instead of doing this in Update, you only need to check if you change textures when the entity takes damage. Override the take damage script and put it in there

Kainkun replied:
This might help you
![image](https://user-images.githubusercontent.com/26811871/109372386-1c2e7e00-785e-11eb-9124-167542cf832e.png)

BUNN1E5 reviewed (changes requested):
Looks mostly good, just a few changes need to be done

Kainkun committed:
Merge branch 'main' into Destructibles

Kainkun committed:
Delete UpgradeLog.htm

Kainkun commented on Assets/Scripts/Destructible.cs:
```diff
@@ -0,0 +1,82 @@
+        HealthPercents[0] = 100;
+        if (numSprites > 1)
+        {
+            float DISTANCE = 100 / (numSprites - 1); //Constant value; Distance from one cutoff to another
```
fyi, usually all caps variables are saved for preprocessor macros. also, C# variable naming convention is camelCase with the first character lowercase. This isn't something we care about, just wanted to let you know

Kainkun committed:
pots!

Kainkun committed:
spelling

Kainkun reviewed (changes requested):
This works exactly like how I hoped it would! Unfortunately, there's a lot of work happening in update that's wasting computation cost. Since this script will be on a lot of objects in the scene, the costs can add up quickly. I'm going to ask that you rework the script to take more advantage of how you extended Entity. Jacob and I put more details in the comments of the code.

Oh yeah, I added some sprites of _beautiful_ pots in Sprites/Dev/Destructible for you to test with :)

j-tomada commented:
Hello, thank you for the feedback! I've been busy as of late trying to care of some midterms coming this Monday. However, whenever I get the chance (probably Monday afternoon), I will get try and resolve some of the issues this script may have.

j-tomada committed:
Made suggested changes

Removed update method; will now change textures only if TakeDamage has been called

Added TakeDamage method (Overrides Entity class)

Changed variable names to C# naming conventions

Stored SpriteRenderer inside a local variable

BUNN1E5 committed:
Moved the OnDeath call to Entity Class

My Bad I forgot to put that in there

BUNN1E5 committed:
Merge branch 'main' into Destructibles

BUNN1E5 reviewed (approved):
Looks Awesome! I moved the death check code from your TakeDamage function into the Entity class btw. Love the Comments too, those can be extremely helpful.

Kainkun committed:
added easybutton to entity take damage and make an example prefab for destructables

Kainkun commented on Assets/Scripts/BaseCode/Entity.cs:
```diff
@@ -10,13 +10,15 @@ public class Entity : MonoBehaviour
     public virtual void TakeDamage(float baseDamage){
         health -= baseDamage;
         if(health <= 0){
-            onDeath();
```
smh, i thought you were a pro @BUNN1E5

BUNN1E5 commented on Assets/Scripts/BaseCode/Entity.cs:
```diff
@@ -10,13 +10,15 @@ public class Entity : MonoBehaviour
         health -= baseDamage;
         if(health <= 0){
-            onDeath();
+            OnDeath();
```
Woopsies, my bad lol

Kainkun reviewed (approved):
awesome! i glad you figured out all the logic to work so well
